src/main.d.ts

```diff
@@ -1,2 +1,5 @@
-declare const fastCartesian: <T>(arrays: T[][]) => T[][]
+declare const fastCartesian: <TFactors extends any[][]>(...arr: TFactors) => {
```

Comment:
This function takes a single argument, not a variadic argument.
(That single argument is an array of arrays)

https://github.com/ehmicky/fast-cartesian/blob/a7cc1e082e48b7b16c46ad726147e167391d0527/src/main.js#L6

Reply:
Ohh I see! I will adjust that, just a moment.

Reply:
That should work right?: https://www.typescriptlang.org/play?target=99&ts=4.0.5#code/CYUwxgNghgTiAEYD2A7AzgF3gMypgwrBiGgJZQoBc8APACoBiUYGSMa8IAHsSsBxQCeAbQC6YgHwAKXCzZpqjZq3YBKeAF4J8AN4AoePGFK5MeKRTwA1iEFJs8EyrSjFTU2mPuVozjxB8HACCMDBQgjQW2CBmdACqKCFhgiDA2gD8jglJ4anw1CggAG4xegC+epW4BEQk5ChSwgZGAIwANPAATB0AzKIdzcIA5HjAQx1DQWhD-YMoAK4QEB3zfCDYFqmzoqpAA

Reply:
@ehmicky I cant get it to work without the rest argument, pretty strange. Currently this will create an union type at 2nd level.

I will have to look at this later again.

Reply:
Thanks for digging into this @Paulomart!

Reply:
@ehmicky figured it out. Sorry for the deplay.

Reply:
Wow, this is some next level typing there!
I tried it out and it works perfectly.

One additional can be also figured out statically would be: when either the top-level array or one of the sub-arrays has length `0`, then the return value is an empty array. However, this one might be harder to type :thinking:

Reply:
> One additional can be also figured out statically would be: when either the top-level array or one of the sub-arrays has length 0, then the return value is an empty array. However, this one might be harder to type 🤔

I think this is still correct, as that is an (empty) tuple definition in that case. 

https://www.typescriptlang.org/play?target=99&ts=4.5.2#code/MYewdgzgLgBAZgQ2gYQQJygUwgSwWGAXhgB4AVAMQWChDQhkwA8swATB-ATwG0BdfgD4AFIhp0IALhg8AdPMrVa9PgEppAbwBQMGYvFoYOAgGtMXEHBj7lEPtJsSejtH0YtM7BgEE0aBFwkxnCYhmQAqmC+-lyYbIIwAPzWkdEBcTDSYJgAbqFaAL78RAnaumiYUACuaARitrIVbFXAmMI6ujDCCAA0MABGqiUwCLJwADYIUACyCAAOwsDD-bIAtvPCbMNy8sB9bG6qqj0dujz8bh1DSCNgXADchVrPWqCQsHNoIM00RPBIUFQGGweDAwnOfD6PAA5AhoX1of14TBocBoXw1I83tAYEw-p9vi0oDwAAwCMkvbEfL4-KAAJj+iBQ6CwuHw4Mxr3AOKYDOIBNpdJ4ABY+M8gA